### PR TITLE
gpu: Enable tile clause in place of collapse with OpenACC

### PR DESCRIPTION
**Pull request by FabioLuporini**
this should generally provide better performance, but there might be exceptions (it would appear 3D elastic being one of them)

### CODE REVIEW

**Review by georgebisbas — COMMENTED** [review]
Maybe add a test with elastic-like structure...?

**Review by mloubout — APPROVED** [review]

**Review by georgebisbas — APPROVED** [review]

File: benchmarks/user/benchmark.py [hunk 1]
```diff
@@ -1,19 +1,13 @@
-from collections import OrderedDict
-import sys
-
 import numpy as np
 import click
 import os
-from devito import (clear_cache, configuration, info, warning, set_log_level,
-                    switchconfig, norm)
+from devito import configuration, info, warning, set_log_level, switchconfig, norm
```
Comment (georgebisbas) [inline benchmarks/user/benchmark.py:4]: ab-order ?

File: devito/passes/iet/languages/openacc.py [hunk 2]
```diff
@@ -151,12 +155,32 @@ def _map_update_wait_device(cls, f, imask=None, queueid=None):
...
 
+    def _make_partree(self, candidates, nthreads=None):
+        assert candidates
+        root = candidates[0]
+
+        if self._is_offloadable(root) and self.par_tile:
+            if isinstance(self.par_tile, tuple):
+                tile = self.par_tile
+            else:
+                # (32,4,4,...) is typically a decent choice
```
Comment (georgebisbas) [inline devito/passes/iet/languages/openacc.py:173]: a decent choice in...V100, RTX3080 etc.. ?

File: tests/test_benchmark.py [hunk 3]
```diff
@@ -8,10 +8,10 @@
 
 
 @pytest.mark.parametrize('mode, problem, op', [
-    ('bench', 'acoustic', 'forward'), ('run', 'acoustic', 'adjoint'),
-    ('run', 'acoustic', 'jacobian'), ('bench', 'acoustic', 'jacobian_adjoint'),
-    ('bench', 'tti', 'forward'), ('bench', 'elastic', 'forward'),
-    ('bench', 'viscoelastic', 'forward'), ('run', 'acoustic_sa', 'forward'),
+    ('run', 'acoustic', 'forward'), ('run', 'acoustic', 'adjoint'),
```
Comment (georgebisbas) [inline tests/test_benchmark.py:11]: ok

File: benchmarks/user/README.md [hunk 4]
```diff
@@ -195,6 +190,12 @@ grid:
 ```
 python benchmark.py run -P tti -d 512 402 890 -so 12 -a basic --tn 100
 ```
+Same as before, but telling devito not to use temporaries to store the
+intermediate values which stem from mixed derivatives:
+```
+python benchmark.py run -P tti -d 512 402 890 -so 12 -a basic --tn 100 --opt
```
Comment (georgebisbas) [inline benchmarks/user/README.md:196]: Good

**Review by georgebisbas — COMMENTED** [review]
Just an idea: Not sure if it is really needed, but why not add some tests to check norms between those two methods. Also we could add some notes for the 32,4,8 tile that this may or may not work in all GPUs. Also, shouldn't this PR include an additional entry (or two entries) in the GPU notebook? Actually no, the GPU notebook is for openmp, but we could show collapse vs tile generated code in one notebook...

File: examples/seismic/utils.py [hunk 5]
```diff
@@ -193,6 +193,23 @@ def __call__(self, parser, args, values, option_string=None):
             values = {'float32': np.float32, 'float64': np.float64}[values]
             setattr(args, self.dest, values)
 
+    class _opt_action(Action):
```
Comment (mloubout) [inline examples/seismic/utils.py:196]: that's duplicate from benchmark, can we put it somewhere common?

Comment (FabioLuporini) [reply examples/seismic/utils.py:196]: they use differ arg parsers so I'm not sure?

Comment (mloubout) [reply examples/seismic/utils.py:196]: Hum slightly different reutrn type indeed, I guess it's fine

**Review by mloubout — APPROVED** [review]

**Pull request merged**